### PR TITLE
Implement functors task #2 - ENikonov

### PR DESCRIPTION
**Pull request by re-nikonov**

### CODE REVIEW

File: functors task (2)/student/include/impl.h [hunk 1]
```diff
@@ -62,10 +59,17 @@ class DataBrowser
...
+        const auto& it = m_dataReaders.find(userId);
+        if (it == m_dataReaders.cend())
+        {
+            return false;
+        }
+        if (it->second == nullptr)
+        {
+            return false;
+        }
+        return std::forward<T>(f)(it->second);
```
Comment (smay1613) [inline functors task (2)/student/include/impl.h:72]: _👍 for forward _

File: functors task (2)/student/src/impl.cpp [hunk 2]
```diff
@@ -5,50 +5,41 @@ void DataBrowser::userLeave(const std::string &userId)
...
-    }
-    if (it->second == nullptr)
-    {
-        return false;
-    }
-    return it->second->getDataType1(returnValues, 0);
+    auto getDataTypeAdapter = [this, &returnValues](const std::unique_ptr<IDataSelector>& selector) {
+        using namespace std::placeholders;
+        return invokeDataRequest(
+            std::bind(&IDataSelector::getDataType1, _1, _2, 0), selector, returnValues);
```
Comment (smay1613) [inline functors task (2)/student/src/impl.cpp:19]: _acceptable, but prefer lambda_

File: functors task (2)/student/src/impl.cpp [hunk 3]
```diff
@@ -5,50 +5,41 @@ void DataBrowser::userLeave(const std::string &userId)
...
-        return false;
-    }
-    if (it->second == nullptr)
-    {
-        return false;
-    }
-    return it->second->getDataType2(returnValues);
+    auto getDataTypeAdapter = [this, &returnValues](const std::unique_ptr<IDataSelector>& selector) {
+        return invokeDataRequest(
+            std::mem_fn(&IDataSelector::getDataType2), selector, returnValues);
```
Comment (smay1613) [inline functors task (2)/student/src/impl.cpp:28]: _lambda is more explicit_

Comment (re-nikonov) [reply functors task (2)/student/src/impl.cpp]: Implemented three different variants to see the difference.

Comment (smay1613) [reply functors task (2)/student/src/impl.cpp]: Nice!

Comment (re-nikonov) [reply functors task (2)/student/src/impl.cpp]: The idea about lambdas is clear. Used std::bind() to see visual difference. Anything else to fix?

Comment (smay1613) [reply functors task (2)/student/src/impl.cpp]: it's ok now

**Pull request closed without merging**